### PR TITLE
Generate better vcd names

### PR DESCRIPTION
**Pull request by ewenmcneill**
Adds a `BaseUsbTestCase` class between `CommonUsbTestCase` and `unittest.TestCase` with a `make_vcd_name` function that generates the full path to the vcd file (rather than repeating that in lots of places).   Attempts to avoid using `__main__` in the file name by replacing it with `__spec__.name` instead (ie, `foo.bar.baz` module name).

So far only `usbcore/sm` has been switched over (other than `usbcore/sm/test.py` which AFAICT isn't runable), but if you're happy with the approach I'll convert over the ones I can find in `usbcore/tx`, `usbcore/rx`, etc.

Ewen

PS: Example filenames created:

```
(LX P=tinyfpga_bx.minimal F=micropython) ewen@parthenon:/src/fpga/litex-buildenv
/third_party/usb-bootloader/valentyusb$ ls vcd
README.md
usbcore.sm.send.TestTxPacketSendAutoCrc.test_ack.vcd
usbcore.sm.send.TestTxPacketSendAutoCrc.test_data0_all_zero.vcd
usbcore.sm.send.TestTxPacketSendAutoCrc.test_data0_descriptor.vcd
usbcore.sm.send.TestTxPacketSendAutoCrc.test_data0_edges.vcd
usbcore.sm.send.TestTxPacketSendAutoCrc.test_data0_one_one.vcd
usbcore.sm.send.TestTxPacketSendAutoCrc.test_data0_one_zero.vcd
usbcore.sm.send.TestTxPacketSendAutoCrc.test_data0_two_ones.vcd
usbcore.sm.send.TestTxPacketSendAutoCrc.test_data1_all_zero.vcd
usbcore.sm.send.TestTxPacketSendAutoCrc.test_nak.vcd
usbcore.sm.send.TestTxPacketSendAutoCrc.test_stall.vcd
usbcore.sm.send.TestTxPacketSendAutoCrc.test_status0.vcd
usbcore.sm.send.TestTxPacketSendAutoCrc.test_status1.vcd
usbcore.sm.send.TestTxPacketSendNoCrc.test_ack.vcd
usbcore.sm.send.TestTxPacketSendNoCrc.test_data0_all_zero.vcd
usbcore.sm.send.TestTxPacketSendNoCrc.test_data0_descriptor.vcd
usbcore.sm.send.TestTxPacketSendNoCrc.test_data0_edges.vcd
usbcore.sm.send.TestTxPacketSendNoCrc.test_data0_one_one.vcd
usbcore.sm.send.TestTxPacketSendNoCrc.test_data0_one_zero.vcd
usbcore.sm.send.TestTxPacketSendNoCrc.test_data0_two_ones.vcd
usbcore.sm.send.TestTxPacketSendNoCrc.test_data1_all_zero.vcd
usbcore.sm.send.TestTxPacketSendNoCrc.test_nak.vcd
usbcore.sm.send.TestTxPacketSendNoCrc.test_stall.vcd
usbcore.sm.send.TestTxPacketSendNoCrc.test_status0.vcd
usbcore.sm.send.TestTxPacketSendNoCrc.test_status1.vcd
(LX P=tinyfpga_bx.minimal F=micropython) ewen@parthenon:/src/fpga/litex-buildenv
/third_party/usb-bootloader/valentyusb$ 
```

### CODE REVIEW

File: valentyusb/usbcore/test/common.py [hunk 1]
```diff
@@ -20,8 +22,29 @@ def grouper(n, iterable, pad=None):
     """
     return zip_longest(*[iter(iterable)]*n, fillvalue=pad)
 
-
-class CommonUsbTestCase(unittest.TestCase):
+# Test case helpers common to all test cases, simple and complex
+#
+class BaseUsbTestCase(unittest.TestCase):
+    # Create a name for the vcd file based on the test case module/class/method
```
Comment (mithro) [inline valentyusb/usbcore/test/common.py]: Convert these comments to docstrings....

File: valentyusb/usbcore/sm/send.py [hunk 2]
```diff
@@ -287,7 +288,7 @@ def stim(dut):
 
         run_simulation(
             dut, stim(dut),
-            vcd_name="vcd/test_token_decode_%s.vcd" % self.id(),
+            vcd_name=self.make_vcd_name(),
```
Comment (mithro) [inline valentyusb/usbcore/sm/send.py:297]: In future I think we add the whole run_simulation call to one of the base classes?

**Review by mithro — COMMENTED** [review]
Approach generally looks great to me. The filenames look great too!

(FYI - Just pushed my latest set of changes which you might want to rebase onto...)

File: valentyusb/usbcore/sm/send.py (hunk 2 above)
Comment (ewenmcneill) [reply valentyusb/usbcore/sm/send.py:297]: Yes, that's my plan if only to hook the process of creating vcd files so that we can build a gtkwave config for it as well.  For the VCD names a simple generator method seemed easier than trying to change it in `**kwargs**` when running the simulation.

Ewen

PS: have swapped comments in test/common.py to docstrings; good suggestion.

**Comment by ewenmcneill** [comment]
@mithro Ready for review again now :-)

Rebased (a couple of hours ago, hopefully close enough).  Updated to include `usbcore/tx` and `usbcore/rx` test changes too.  Also updated `usbcore/cpu` test changes, but AFAICT none of those currently work before/after change (seemingly due to other code changes from a quick glance).

This leaves:

```
(LX P=tinyfpga_bx.minimal F=micropython) ewen@parthenon:/src/fpga/litex-buildenv
/third_party/usb-bootloader/valentyusb$ find usbcore -type f | xargs grep vcd/ |
 grep -v __pycache
usbcore/sm/test.py:            vcd_name="vcd/%s.vcd" % self.id(),
usbcore/test/common.py:        return ("vcd/%s.vcd" % basename)
(LX P=tinyfpga_bx.minimal F=micropython) ewen@parthenon:/src/fpga/litex-buildenv
/third_party/usb-bootloader/valentyusb$ 
```

generating `vcd` filename paths. 

`usbcore/sm/test.py` appears to be fragment of code without includes, that cannot be run by itself.  Possibly that one is clutter to be removed?  Or needs to be made to work somehow.  I've left it alone as I'm unclear what its status was.

Ewen

File: valentyusb/usbcore/test/common.py (hunk 1 above)
Comment (ewenmcneill) [reply valentyusb/usbcore/test/common.py]: Converted.

File: valentyusb/usbcore/tx/tester.py [hunk 3]
```diff
@@ -1,12 +1,24 @@
 #!/usr/bin/env python3
 
+import inspect
+
 from migen import *
 
 MIGEN_SIGNALS = ("reset", "ce")
 
+# Helper to find the ultimate caller's module name (extra level further up
```
Comment (mithro) [inline valentyusb/usbcore/tx/tester.py]: Would be good if this was a docstring.

File: valentyusb/usbcore/test/common.py [hunk 4]
```diff
@@ -20,8 +21,33 @@ def grouper(n, iterable, pad=None):
     """
     return zip_longest(*[iter(iterable)]*n, fillvalue=pad)
 
-
-class CommonUsbTestCase(unittest.TestCase):
+class BaseUsbTestCase(unittest.TestCase):
+    """
+    Test case helpers common to all test cases, simple and complex
+    """
+    def make_vcd_name(self, basename=None, modulename=None):
```
Comment (mithro) [inline valentyusb/usbcore/test/common.py]: Rather then supply basename, could we instead supply an "extra" component that way we don't have to duplicate the basename values in these other tests?

Comment (ewenmcneill) [reply valentyusb/usbcore/test/common.py]: Good idea.  I've added a `testsuffix=` optional parameter, which gets appended to the (usually auto-derived) basename.  And replaced all the `basename=...` overrides with that.  I've left `basename` as a parameter override in case we need it later (overrides `self.id()`) since it's a tiny bit of code.  But currently only the `modulename=` override is used directly.

This updates the names of a number of the vcd output files again (to include class names / method names too), but the predictability of naming is probably useful anyway.

Ewen

File: valentyusb/usbcore/tx/tester.py (hunk 3 above)
Comment (ewenmcneill) [reply valentyusb/usbcore/tx/tester.py]: Done.

**Comment by ewenmcneill** [comment]
@mithro I think that covers all the changes you wanted.  Possibly worth merging soonish so that it doesn't need rebasing again, as this pull request inherently touches a lot of files... :-)

Ewen

**Comment by mithro** [comment]
Github is reporting a conflict in `valentyusb/usbcore/sm/send.py`? Rebase and I'll merge...

**Comment by ewenmcneill** [comment]
@mithro Rebased again.   AFAICT the only conflict was:

```
-class CommonTxPacketSendTestCase:
+class CommonTxPacketSendTestCase(BaseUsbTestCase):
```

which it looks like maybe you changed that line too, but the base class (to pick up the shared behaviour) should in theory be safe (and tests seems to run as much as they did before).

Ewen

**Comment by mithro** [comment]
Merged!
